File: convertCSVInputsToReferentiels.js
import fs from 'fs'
import path from 'path'

import { createReadStream, writeFileSync } from "fs";
import Papa from "papaparse";

import _ from 'lodash'

import departements from './data/departements.json' assert {type: 'json'}

import { createClient } from '@supabase/supabase-js'
// DEV
// 'https://drncrjteathtblggsgxi.supabase.co'
// PROD
// 'https://ixxbyuandbmplfnqtxyw.supabase.co'
const supabase = createClient('https://ixxbyuandbmplfnqtxyw.supabase.co', process.env.SUPABASE_ADMIN_KEY, {
  auth: { persistSession: false }
})

function readJsonFilesInFolder(folderPath, cb) {
  try {
    // Read all files in the folder
    const files = fs.readdirSync(folderPath);

    // Filter for JSON files
    const jsonFiles = files.filter(file => file.endsWith('.json'));

    // Process each JSON file
    jsonFiles.forEach(file => {
      const filePath = path.join(folderPath, file);
      
      // Read the content of the JSON file
      const rawData = fs.readFileSync(filePath, 'utf8');
      
      // Parse the JSON data
      const jsonData = JSON.parse(rawData);
      
      console.log(`Processing file: ${file}`);
      // console.log('Data:', jsonData);

      cb(jsonData)

      // Here you can manipulate or process the jsonData as needed
      // For example: 
      // jsonData.forEach(item => { console.log(item); });
    });

  } catch (err) {
    console.error('Error reading JSON files:', err);
  }
}

async function parseCsv(filePath) {
  console.log(`Starting to parse: ${filePath}`); // Log when parsing starts

  return await new Promise((resolve) => {
    const stream = createReadStream(filePath);

    Papa.parse(stream, {
      header: true,
      error: (err) => {
        console.log(`Error parsing ${filePath}:`, err); // Log parsing errors
      },
      complete: (result) => {
        console.log(`Finished parsing: ${filePath}`); // Log when parsing finishes
        resolve(result);
      }
    });
  });
}

async function parseInputs () {
  try {
    console.time('CSV Parsing Time'); // Measure total time for all parsing operations

    const { data: communeSiren } = await parseCsv('./inputs/csv/communes_siren.csv');
    const { data: communes } = await parseCsv('./inputs/csv/communes_2024.csv');
    const { data: compositionCommunes } = await parseCsv('./inputs/csv/Composition_communale-Table 1.csv');
    const { data: compositionsEPT } = await parseCsv('./inputs/csv/Composition_communale-Table_EPT.csv');

    console.timeEnd('CSV Parsing Time'); // Measure total time for all parsing operations
    console.log('All CSV files parsed successfully.', communeSiren.length, communes.length);

    return {communeSiren, communes, compositionCommunes, compositionsEPT}
  } catch (error) {
    console.error('Error during CSV parsing:', error); // Catch and log any unexpected errors
  }
}

const {
  communeSiren,
  communes,
  compositionCommunes,
  compositionsEPT
} = await parseInputs()

const communesSirenMap = {}
const communesInseeMap = {}
const groupementsMap = {}

communeSiren.forEach(commune => {
  const code = commune['Code INSEE de la commune']
  const siren = commune['Siren']

  const inseeCommune = communes.find(c => c.COM === code)

  communesSirenMap[siren] = {
    code,
    siren,
    type: inseeCommune.TYPECOM,
    intitule: inseeCommune.LIBELLE,
    regionCode: inseeCommune.REG,
    departementCode: inseeCommune.DEP,
    groupements: [],
    membres: [],
    intercommunaliteCode: '',
    competencePLU: true,
    competenceSCOT: false
  }

  communesInseeMap[code] = communesSirenMap[siren]
})

compositionCommunes.forEach(commune => {
  const insee = commune.CODGEO
  const epci = commune.EPCI

  if(epci && !epci.includes('Z')) {
    communesInseeMap[insee].intercommunaliteCode = epci
  }
})

compositionsEPT.forEach(commune => {
  const insee = commune.CODGEO
  const epci = commune.EPCI

  if(epci && !epci.includes('Z')) {
    communesInseeMap[insee].intercommunaliteCode = epci
  }
})

let nbMissingCommunes = 0

communes.forEach(commune => {
  if(commune.TYPECOM === 'COMD' || commune.TYPECOM === 'COMA') {
    const parent = communesInseeMap[commune.COMPARENT]

    communesInseeMap[`${commune.COM}_COMD`] = {
      code: commune.COM,
      codeParent: commune.COMPARENT,
      siren: '',
      type: commune.TYPECOM,
      intitule: commune.LIBELLE,
      regionCode: commune.REG,
      departementCode: commune.DEP,
      groupements: [],
      membres: [],
      intercommunaliteCode: '',
      competencePLU: false,
      competenceSCOT: false
    }

    parent.membres.push({
      code: commune.COM,
      type: commune.TYPECOM,
      intitule: commune.LIBELLE,
    })
  } else {
    // We check if all communes are inserted from Banatic
    if(!communesInseeMap[commune.COM]) {
      nbMissingCommunes += 1
      console.log('missing commune in Banatic', commune.COM)
    }
  }
})

// We iterate a first time to populate all groupements
readJsonFilesInFolder('./inputs/json', (membres) => {
  membres.forEach(membre => {
    const sirenGroupement = membre['N° SIREN']
    const departement = departements.find(d => d.intitule === membre['Département'])

    if(!groupementsMap[sirenGroupement]) {
      groupementsMap[sirenGroupement] = {
        type: membre['Nature juridique'],
        intitule: membre['Nom du groupement'],
        siren: sirenGroupement,
        code: sirenGroupement,
        regionCode: departement.region.code,
        departementCode: departement.code,
        competencePLU: membre["5510 - Plan local d'urbanisme et document d'urbanisme en tenant lieu (Art. L. 153-1 du code de l'urbanisme)"] === 'OUI',
        competenceSCOT: membre["5500 - Schéma de cohérence territoriale (SCOT) (Art. L. 143-16 code de l'urbanisme)"] === 'OUI',
        membres: [],
        groupements: []
      }
    }
  })
})

let missingMembres = 0

// Then a second time to populate all members.
readJsonFilesInFolder('./inputs/json', (membres) => {
  // We iterate region by region to create group referentiel.
  membres.forEach(membre => {
    const sirenGroupement = membre['N° SIREN']
    const sirenMembre = membre['Siren membre']

    const groupement = groupementsMap[sirenGroupement]
    const membreRef = communesSirenMap[sirenMembre] || groupementsMap[sirenMembre]

    if(membreRef) {
      membreRef.groupements.push({
        type: membre['Nature juridique'],
        intitule: membre['Nom du groupement'],
        siren: sirenGroupement,
        code: sirenGroupement,
      })

      if(groupement.competencePLU) {
        membreRef.competencePLU = false
      }

      groupement.membres.push({
        type: membreRef.type,
        intitule: membreRef.intitule,
        siren: sirenMembre,
        code: membreRef.code
      })
    } else {
      missingMembres += 1
      // console.log('Missing member', sirenMembre)
    }
  })
})

console.log(Object.keys(groupementsMap).length)
console.log(Object.keys(communesInseeMap).length)
console.log(`Missing communes: ${nbMissingCommunes}, missing membres: ${missingMembres}`)

const groupementsArray = _.map(groupementsMap, group => group)
const communesArray = _.map(communesInseeMap, c => c)

const searchableGroupements = []

for (let index = 0; index < groupementsArray.length; index++) {
  const group = groupementsArray[index];

  if(group.competencePLU || group.competenceSCOT) {
    searchableGroupements.push(group)
  } else {
    const {data} = await supabase.from('procedures').select('id')
      .in('status', ['opposable', 'en cours'])
      .eq('collectivite_porteuse_id', group.code)
      .limit(1)

    if(data.length) {
      searchableGroupements.push(group)
    }
  }

  console.log(`search filter ${index}/${groupementsArray.length}`)
}

console.log('searchableGroupements', searchableGroupements.length)

fs.writeFileSync('./output/groupements_2024_map.json', JSON.stringify(groupementsMap, null, 2))
fs.writeFileSync('./output/communes_2024_map.json', JSON.stringify(communesInseeMap, null, 2))

fs.writeFileSync('./output/groupements_2024.json', JSON.stringify(groupementsArray, null, 2))
fs.writeFileSync('./output/communes_2024.json', JSON.stringify(communesArray, null, 2))

fs.writeFileSync('./output/groupements_competents_2024.json', JSON.stringify(searchableGroupements, null, 2))

File: convert_csv_inputs_to_referentiels.py
from csv import DictReader
from functools import partial
from itertools import chain
from json import dumps, load
from operator import attrgetter, itemgetter
from pathlib import Path

# FIXME mise dependencies + in2csv ?
departements_by_intitule = {
    departement["intitule"]: departement
    for departement in load(Path("./data/departements.json").open())
}

csv = Path("inputs/csv")
input_communes_banatic = list(DictReader((csv / "communes_siren.csv").open()))

input_communes_insee = list(DictReader((csv / "communes_2024.csv").open()))
input_communes_insee_by_code_insee = {
    commune["COM"]: commune
    for commune in input_communes_insee
    if commune["TYPECOM"] == "COM"
}


print(
    f"All CSV files parsed successfully. {len(input_communes_banatic)} {len(input_communes_insee)}"
)


def format_commune(commune_banatic):
    code_commune = commune_banatic["Code INSEE de la commune"]
    commune_insee = input_communes_insee_by_code_insee[code_commune]
    return {
        "code": code_commune,
        "siren": commune_banatic["Siren"],
        "type": commune_insee["TYPECOM"],
        "intitule": commune_insee["LIBELLE"],
        "regionCode": commune_insee["REG"],
        "departementCode": commune_insee["DEP"],
        "groupements": [],
        "membres": [],
        "intercommunaliteCode": "",
        "competencePLU": True,
        "competenceSCOT": False,
    }


def membres_de_groupement():
    for f in sorted(Path("./inputs/json").glob("*.json"), key=attrgetter("name")):
        print(f"Processing file: {f.name}")
        yield from load(f.open())


commune_by_code_insee = {}
commune_by_siren = {}
for commune_banatic in input_communes_banatic:
    commune = format_commune(commune_banatic)
    commune_by_code_insee[commune["code"]] = commune
    commune_by_siren[commune["siren"]] = commune

compositions_EPCI = DictReader((csv / "Composition_communale-Table 1.csv").open())
compositions_EPT = DictReader((csv / "Composition_communale-Table_EPT.csv").open())
for commune in chain(compositions_EPCI, compositions_EPT):
    insee = commune["CODGEO"]
    epci = commune.get("EPCI")
    if epci and "Z" not in epci:
        commune_by_code_insee[insee]["intercommunaliteCode"] = epci

missing_communes = []
for commune in input_communes_insee:
    if commune["TYPECOM"] not in ["COMD", "COMA"]:
        if commune["COM"] not in commune_by_code_insee:
            missing_communes.append(commune)
            print(f"missing commune in Banatic {commune['COM']}")
        continue

    commune_by_code_insee[commune["COM"] + "_COMD"] = {
        "code": commune["COM"],
        "codeParent": commune["COMPARENT"],
        "siren": "",
        "type": commune["TYPECOM"],
        "intitule": commune["LIBELLE"],
        "regionCode": commune["REG"],
        "departementCode": commune["DEP"],
        "groupements": [],
        "membres": [],
        "intercommunaliteCode": "",
        "competencePLU": False,
        "competenceSCOT": False,
    }

    parent = commune_by_code_insee[commune["COMPARENT"]]
    parent["membres"].append(
        {
            "code": commune["COM"],
            "type": commune["TYPECOM"],
            "intitule": commune["LIBELLE"],
        }
    )

groupements_by_siren = {}
for membre in membres_de_groupement():
    siren_groupement = membre["N° SIREN"]
    if siren_groupement in groupements_by_siren:
        continue

    departement = departements_by_intitule[membre["Département"]]
    groupements_by_siren[siren_groupement] = {
        "type": membre["Nature juridique"],
        "intitule": membre["Nom du groupement"],
        "siren": siren_groupement,
        "code": siren_groupement,
        "regionCode": departement["region"]["code"],
        "departementCode": departement["code"],
        "competencePLU": "OUI"
        == membre[
            "5510 - Plan local d'urbanisme et document d'urbanisme en tenant lieu (Art. L. 153-1 du code de l'urbanisme)"
        ],
        "competenceSCOT": "OUI"
        == membre[
            "5500 - Schéma de cohérence territoriale (SCOT) (Art. L. 143-16 code de l'urbanisme)"
        ],
        "membres": [],
        "groupements": [],
    }


missing_membres = []
for membre in membres_de_groupement():
    siren_mem = membre["Siren membre"]
    membre_obj = commune_by_siren.get(siren_mem) or groupements_by_siren.get(siren_mem)
    if not membre_obj:
        missing_membres.append(siren_mem)
        continue

    siren_groupement = membre["N° SIREN"]
    groupement = groupements_by_siren[siren_groupement]

    membre_obj["groupements"].append(
        {
            "type": membre["Nature juridique"],
            "intitule": membre["Nom du groupement"],
            "siren": siren_groupement,
            "code": siren_groupement,
        }
    )

    if groupement["competencePLU"]:
        membre_obj["competencePLU"] = False

    groupement["membres"].append(
        {
            "type": membre_obj["type"],
            "intitule": membre_obj["intitule"],
            "siren": siren_mem,
            "code": membre_obj["code"],
        }
    )


print(f"groupementsMap {len(groupements_by_siren)}")
print(f"communesInseeMap {len(commune_by_code_insee)}")
print(
    f"Missing communes: {len(missing_communes)}, missing membres: {len(missing_membres)}"
)

searchable_groupements = [
    groupement
    for groupement in groupements_by_siren.values()
    if groupement["competencePLU"] or groupement["competenceSCOT"]
]
print(f"searchableGroupements {len(searchable_groupements)}")

jason = partial(dumps, indent=2, ensure_ascii=False)

Path("./output/communes_2024_map.json").write_text(jason(commune_by_code_insee))
Path("./output/communes_2024.json").write_text(
    jason(sorted(commune_by_code_insee.values(), key=itemgetter("code")))
)


Path("./output/groupements_2024.json").write_text(
    jason(sorted(groupements_by_siren.values(), key=itemgetter("code")))
)
Path("./output/groupements_2024_map.json").write_text(jason(groupements_by_siren))
Path("./output/groupements_competents_2024.json").write_text(
    jason(sorted(searchable_groupements, key=itemgetter("code")))
)
